cdap-common/src/main/resources/cdap-default.xml

```diff
@@ -356,9 +356,9 @@
 
   <property>
     <name>app.program.jvm.opts</name>
-    <value>-XX:MaxPermSize=128M ${twill.jvm.gc.opts}</value>
+    <value>-XX:MaxPermSize=128M -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=&lt;LOG_DIR&gt; ${twill.jvm.gc.opts}</value>
```

Comment:
Better separate out as a different property, something like `app.program.container.oom.opts` and use `${app.program.container.oom.opts}` here. It's easy to turn it  on/off in that way.